builtin/logical/transit/backend.go

```diff
@@ -231,29 +231,36 @@ func (b *backend) autoRotateKeys(ctx context.Context, req *logical.Request) erro
 			continue
 		}
 
-		// If the policy's automatic rotation interval is 0, it should not
-		// automatically rotate.
-		if p.AutoRotateInterval == 0 {
-			continue
-		}
-
-		// Retrieve the latest version of the policy and determine if it is time to rotate.
-		latestKey := p.Keys[strconv.Itoa(p.LatestVersion)]
-		if time.Now().After(latestKey.CreationTime.Add(p.AutoRotateInterval)) {
-			if b.Logger().IsDebug() {
-				b.Logger().Debug("automatically rotating key", "key", key)
-			}
-			if !b.System().CachingDisabled() {
-				p.Lock(true)
-			}
-			err = p.Rotate(ctx, req.Storage, b.GetRandomReader())
-			p.Unlock()
-			if err != nil {
-				errs = multierror.Append(errs, err)
-				continue
-			}
+		err = b.rotateIfRequired(ctx, req, key, p)
+		if err != nil {
+			errs = multierror.Append(errs, err)
 		}
 	}
 
 	return errs.ErrorOrNil()
 }
+
+// rotateIfRequired rotates a key if it is due for autorotation.
+func (b *backend) rotateIfRequired(ctx context.Context, req *logical.Request, key string, p *keysutil.Policy) error {
+	if !b.System().CachingDisabled() {
+		p.Lock(true)
+	}
+	defer p.Unlock()
```

Comment:
Is it safe to unlock if we didn't just lock?

Reply:
Oh, i see:

> Calls to LockManager.GetPolicy will return a locked policy when caching is disabled, which is why calls to Policy.Unlock() won't panic when caching is enabled

Reply:
So you are correct Brian, it is not safe to call Unlock if you don't have a previous lock but in this case we do have the lock as Matt mentioned in the summary.

It's an odd pattern and I had to prove it to myself, the lock that is grabbed within the lock manager at line [293](https://github.com/hashicorp/vault/blob/main/sdk/helper/keysutil/lock_manager.go#L293) is not released if caching is disabled (look at the deferred cleanup function a bit lower). 

So we need to perform this pattern or else we will be grabbing a double lock if caching is disabled and hang, but if caching was enabled, the lock within the lock manager was released so we need to grab it again...